### PR TITLE
tests(execution): runtime decoder

### PR DESCRIPTION
**Pull request by nicksinch**
**Detailed description**:
- Adds unit tests for the runtime decoder

**Which issue(s) this PR fixes**:
Part of [#215](https://github.com/orgs/LimeChain/projects/5/views/1?pane=issue&itemId=40667258)

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: execution/types/runtime_decoder_test.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+	primitives "github.com/LimeChain/gosemble/primitives/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+var (
+	mockModuleOne        *mocks.Module
+	mockCallOne          *mocks.Call
+	moduleOneIdx         = sc.U8(0)
+	moduleIdx            = sc.U8(0)
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]: do you need both?

File: execution/types/runtime_decoder_test.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+	functionIdx          = sc.U8(0)
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]: not super important, but we usually tend to separate declarations of constants, variables, ...

```
const (
	moduleOneIdx = sc.U8(0)
	moduleIdx    = sc.U8(0)
	functionIdx = sc.U8(0)
)


var (
	signedExtrinsicBytes = []byte{
   ...
)

var (
	mockModuleOne *mocks.Module
	mockCallOne   *mocks.Call
)

File: execution/types/runtime_decoder_test.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+	expectedBlock := NewBlock(header, sc.Sequence[primitives.UncheckedExtrinsic]{})
+
+	assert.Equal(t, resultBlock, expectedBlock)
+}
+
+func Test_RuntimeDecoder_DecodeBlock_Module_Not_Exists(t *testing.T) {
+	target := setupRuntimeDecoder()
+
+	moduleFunctions[0] = mockCallOne
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]: I think you don't need this

File: execution/types/runtime_decoder_test.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+		uint8(idxModuleNotExists), uint8(functionIdx), // call
+	}
+
+	buffExtrinsic := bytes.NewBuffer(append(sc.ToCompact(len(nonExistentExtrinsicsBytes)).Bytes(), nonExistentExtrinsicsBytes...))
+
+	lenExtrinsics := sc.ToCompact(1).Bytes()
+	decodeBlockBytes := append(lenExtrinsics, buffExtrinsic.Bytes()...)
+	decodeBlockBytes = append(header.Bytes(), decodeBlockBytes...)
+
+	mockCallOne.On("Encode", mock.Anything)
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]: I think you don't need this

File: execution/types/runtime_decoder_test.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
+	mockSignedExtra.On("Decode", mock.Anything).Return()
+	mockModuleOne.On("Functions").Return(moduleFunctions)
+	mockCallOne.On("DecodeArgs", decodeBlockBuff).Return(mockCallOne)
+	result := target.DecodeBlock(decodeBlockBuff)
+
+	mockCallOne.On("DecodeArgs", buffExtrinsic).Return(mockCallOne)
+
+	decodedExtr := target.DecodeUncheckedExtrinsic(buffExtrinsic)
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]: I don't think you should be calling this method manually, as it is not a mock and is being called as part of the DecodeBlock, which is what we are testing here.

Also, you may want to add a separate test for the DecodeUncheckedExtrinsic method which is public and used in other places

```go
func Test_RuntimeDecoder_DecodeUncheckedExtrinsic(t *testing.T) {
	target := setupRuntimeDecoder()

	moduleFunctions[0] = mockCallOne

	buf := bytes.NewBuffer(
		append(sc.ToCompact(len(signedExtrinsicBytes)).Bytes(), signedExtrinsicBytes...),
	)

	mockSignedExtra.On("Decode", mock.Anything).Return()
	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
	mockModuleOne.On("Functions").Return(moduleFunctions)
	mockCallOne.On("DecodeArgs", buf).Return(mockCallOne)

	decodedExtr := target.DecodeUncheckedExtrinsic(buf)

	mockSignedExtra.AssertCalled(t, "Decode", mock.Anything)
	mockModuleOne.AssertCalled(t, "GetIndex")
	mockModuleOne.AssertCalled(t, "Functions")
	mockCallOne.AssertCalled(t, "DecodeArgs", buf)
	assert.Equal(t, decodedExtr.IsSigned(), true)
}
```

Comment (radkomih) [reply execution/types/runtime_decoder_test.go]: Should be something like this

```go
func Test_RuntimeDecoder_DecodeBlock_Single_Extrinsic(t *testing.T) {
	target := setupRuntimeDecoder()

	moduleFunctions[0] = mockCallOne

	buffExtrinsic := bytes.NewBuffer(append(sc.ToCompact(len(signedExtrinsicBytes)).Bytes(), signedExtrinsicBytes...))

	lenExtrinsics := sc.ToCompact(1).Bytes()
	decodeBlockBytes := append(lenExtrinsics, buffExtrinsic.Bytes()...)
	decodeBlockBytes = append(header.Bytes(), decodeBlockBytes...)

	decodeBlockBuff := bytes.NewBuffer(decodeBlockBytes)

	mockSignedExtra.On("Decode", mock.Anything).Return()
	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
	mockModuleOne.On("Functions").Return(moduleFunctions)
	mockCallOne.On("DecodeArgs", decodeBlockBuff).Return(mockCallOne)

	result := target.DecodeBlock(decodeBlockBuff)
```

You can check the `unchecked_extrinsic_test.go` file to see how to initialize an extrinsic and use its encoding for the signedExtrinsicBytes.

```go
	extrinsics := sc.Sequence[primitives.UncheckedExtrinsic]{
		NewUncheckedExtrinsic( ... ),
	}
```
```go
	expectedBlock := NewBlock(header, extrinsics)

	assert.Equal(t, expectedBlock.Header(), result.Header())
	assert.Equal(t, expectedBlock.Extrinsics(), result.Extrinsics())
```
Once it is decoded, assert the fields/methods. Calling `.Bytes()` encodes it again.

File: execution/types/runtime_decoder_test.go [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	sc "github.com/LimeChain/goscale"
+	"github.com/LimeChain/gosemble/mocks"
+	primitives "github.com/LimeChain/gosemble/primitives/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+var (
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]: `const` instead of `var`

File: execution/types/runtime_decoder_test.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+
+	extrinsics := sc.Sequence[primitives.UncheckedExtrinsic]{
+		NewUncheckedExtrinsic(sc.U8(signedExtrinsicVersion), extrinsicSignature, mockCallOne, mockSignedExtra),
+	}
+
+	expectedBlock := NewBlock(header, extrinsics)
+	assert.Equal(t, expectedBlock.Header(), result.Header())
+	mockCallOne.On("Encode", mock.Anything)
+	mockSignedExtra.On("Encode", mock.Anything)
+	assert.Equal(t, expectedBlock.Extrinsics().Bytes(), result.Extrinsics().Bytes())
```
Comment (radkomih) [inline execution/types/runtime_decoder_test.go]:
```go
assert.Equal(t,
  block.Extrinsics()[0].Signature(),
  result.Extrinsics()[0].Signature(),
)

assert.Equal(t,
  block.Extrinsics()[0].Function(),
  result.Extrinsics()[0].Function(),
)

assert.Equal(t,
  block.Extrinsics()[0].Extra(),
  result.Extrinsics()[0].Extra(),
)

assert.Equal(t,
  block.Extrinsics()[0].IsSigned(),
  result.Extrinsics()[0].IsSigned(),
)
```

**Review by radkomih — DISMISSED** [review]

**Comment by failfmi** [comment]
There are two missing test cases with DecodeUncheckedExtrinsic - `invalid extrinsic version` and `invalid length prefix`

File: execution/types/runtime_decoder_test.go [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+
+	mockModuleOne.On("GetIndex").Return(moduleOneIdx)
+	mockSignedExtra.On("Decode", mock.Anything).Return()
+
+	assert.PanicsWithValue(t, "module with index ["+strconv.Itoa(int(idxModuleNotExists))+"] not found", func() {
+		target.DecodeBlock(decodeBlockBuff)
+	})
+}
+
+func Test_RuntimeDecoder_DecodeBlock_Function_Not_Exists(t *testing.T) {
```
Comment (failfmi) [inline execution/types/runtime_decoder_test.go]: I think that this test has to be for DecodeCall.

File: execution/types/runtime_decoder_test.go [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+	buff := bytes.NewBuffer(append(header.Bytes(), lenExtrinsics...))
+
+	resultBlock := target.DecodeBlock(buff)
+
+	expectedBlock := NewBlock(header, sc.Sequence[primitives.UncheckedExtrinsic]{})
+
+	assert.Equal(t, expectedBlock, resultBlock)
+}
+
+func Test_RuntimeDecoder_DecodeBlock_Module_Not_Exists(t *testing.T) {
```
Comment (failfmi) [inline execution/types/runtime_decoder_test.go]: I think that this test has to be for DecodeCall.

**Review by failfmi — DISMISSED** [review]

File: execution/types/runtime_decoder_test.go (hunk 8 above)
Comment (nicksinch) [reply execution/types/runtime_decoder_test.go]: Fixed

File: execution/types/runtime_decoder_test.go (hunk 9 above)
Comment (nicksinch) [reply execution/types/runtime_decoder_test.go]: Fixed

**Review by failfmi — APPROVED** [review]

**Review by radkomih — APPROVED** [review]

**Pull request merged**